Bottom: d72f7db240fa874512ae9ef17fcb18a2808e7116
Top:    ad501370940f9b71f8fa4429ce3a662b68af20b5
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-11-03 02:23:24 +0000

Play with perspective


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index 79619aa..b636e47 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -1421,7 +1421,8 @@ DrawMask (BoxType * screen)
   polygon.Clipped = board_outline_poly ();
   polygon.NoHoles = NULL;
   polygon.NoHolesValid = 0;
-  polygon.BoundingBox = *screen;
+  if (screen)
+    polygon.BoundingBox = *screen;
   SET_FLAG (FULLPOLYFLAG, &polygon);
   common_fill_pcb_polygon (out->fgGC, &polygon, screen);
   poly_Free (&polygon.Clipped);
@@ -1853,20 +1854,37 @@ ghid_drawing_area_expose_cb (GtkWidget *widget,
 
   glMatrixMode (GL_PROJECTION);
   glLoadIdentity ();
-  glOrtho (0, widget->allocation.width, widget->allocation.height, 0, -100000, 100000);
+#ifdef VIEW_ORTHO
+  glOrtho (-1, 1, 1, -1, 10, 12);
+#else
+  glFrustum (-1, 1, 1, -1, 1, 24);
+#endif
   glMatrixMode (GL_MODELVIEW);
   glLoadIdentity ();
 
-  glTranslatef (widget->allocation.width / 2., widget->allocation.height / 2., 0);
+  /* TEST HACK */
+  glScalef (10., 10., 1.);
+
+  /* Push the space coordinates board back into the middle of the z-view volume */
+  glTranslatef (0., 0., -11.);
+
+  /* Rotate about the center of the board space */
   glMultMatrixf ((GLfloat *)view_matrix);
-  glTranslatef (-widget->allocation.width / 2., -widget->allocation.height / 2., 0);
-  glScalef ((ghid_flip_x ? -1. : 1.) / port->zoom,
-            (ghid_flip_y ? -1. : 1.) / port->zoom,
-            ((ghid_flip_x == ghid_flip_y) ? 1. : -1.) / port->zoom);
-  glTranslatef (ghid_flip_x ? port->view_x0 - PCB->MaxWidth  :
-                             -port->view_x0,
-                ghid_flip_y ? port->view_y0 - PCB->MaxHeight :
-                             -port->view_y0, 0);
+
+  /* Flip about the center of the viewed area */
+  glScalef ((ghid_flip_x ? -1. : 1.), (ghid_flip_y ? -1. : 1.), ((ghid_flip_x == ghid_flip_y) ? 1. : -1.));
+
+  /* Scale board coordiantes to (-1,-1)-(1,1) coordiantes */
+  glScalef (2. / port->zoom / widget->allocation.width,
+            2. / port->zoom / widget->allocation.height,
+            2. / port->zoom / widget->allocation.width); /* Not sure about the z-scaling */
+
+  /* Translate to the center of the board space view */
+  glTranslatef (ghid_flip_x ? port->view_x0 - PCB->MaxWidth / 2  : port->view_x0 - PCB->MaxWidth / 2  ,
+                ghid_flip_y ? port->view_y0 - PCB->MaxHeight / 2 : port->view_y0 - PCB->MaxHeight / 2 ,
+                0);
+
+  /* Stash the model view matrix so we can work out the screen coordinate -> board coordinate mapping */
   glGetFloatv (GL_MODELVIEW_MATRIX, (GLfloat *)last_modelview_matrix);
 
 #ifdef ONE_SHOT
@@ -2010,7 +2028,8 @@ ghid_drawing_area_expose_cb (GtkWidget *widget,
 
 
   // hid_expose_callback (&ghid_hid, &region, 0);
-  ghid_draw_everything (&region);
+  // ghid_draw_everything (&region);
+  ghid_draw_everything (NULL);
   hidgl_flush_triangles (&buffer);
 
   /* Just prod the drawing code so the current depth gets set to
